educator/src/Dscp/DB/SQL/Util/Json.hs

```diff
@@ -0,0 +1,172 @@
+    -- | Type of the accessed field.
+    type GJsonFieldType name x :: *
+
+    -- | Get name of the field as it is encoded in JSON.
+    gJsonFieldName :: Aeson.Options -> Text
+
+instance TypeError ('Text "Lol no, sum types are not supported") =>
+         GHasJsonField orig name (G.D1 dm (c1 G.:+: c2)) where
+
+    type GJsonFieldType name (G.D1 dm (c1 G.:+: c2)) =
+        TypeError ('Text "Cannot access sum type")
+
+    gJsonFieldName =
+        error "no way"
+
+instance TypeError ('Text "Constructor has no fields") =>
+         GHasJsonField orig name (G.D1 dm (G.C1 cm G.U1)) where
+
+    type GJsonFieldType name (G.D1 dm (G.C1 cm G.U1)) =
+        TypeError ('Text "Constructor has no fields")
+
+    gJsonFieldName =
+        error "nothing is here"
+
+instance TypeError ('Text "Accessing constructors with one field \
+                          \is not yet implemented") =>
+         GHasJsonField orig name (G.D1 dm (G.C1 cm (G.S1 sm (G.Rec0 inner)))) where
+
+    type GJsonFieldType name (G.D1 dm (G.C1 cm (G.S1 sm (G.Rec0 inner)))) =
+        TypeError ('Text "Cannot yet access constructor with only one field")
+
+    gJsonFieldName =
+        -- Rules for deriving JSON field in this case are a bit
+        -- more complex than usual.
+        error "maybe some day..."
+
+instance ( KnownSymbol name
+         , GProductHasJsonField orig name (s1 G.:*: s2)
+         ) =>
+         GHasJsonField orig name (G.D1 dm (G.C1 cm (s1 G.:*: s2))) where
+
+    type GJsonFieldType name (G.D1 dm (G.C1 cm (s1 G.:*: s2))) =
+        GProductJsonFieldType name (s1 G.:*: s2)
+
+    gJsonFieldName opts =
+        toText $ Aeson.fieldLabelModifier opts (symbolVal (Proxy @name))
+
+-- | Access a JSON field in product type.
+type family GProductJsonFieldType name (x :: * -> *) :: * where
+
+    GProductJsonFieldType name (x G.:*: y) =
+        If (GProductJsonFieldType name x == FieldNotFound)
+           (GProductJsonFieldType name y)
+           (GProductJsonFieldType name x)
+
+    GProductJsonFieldType name (G.S1 ('G.MetaSel ('Just fname) u s l) (G.Rec0 ty))
+        = If (name == fname) ty FieldNotFound
+
+    GProductJsonFieldType name (G.S1 ('G.MetaSel 'Nothing u s l) (G.Rec0 ty))
+        = FieldNotFound
+
+-- | Whether a field is present in a type product.
+type family GProductHasJsonField orig name x :: Constraint where
+    GProductHasJsonField orig name x =
+        If (GProductJsonFieldType name x == FieldNotFound)
+            (TypeError ('Text "No field " ':<>: 'ShowType name ':<>:
+                        'Text " in " ':<>: 'ShowType orig))
+            (() :: Constraint)
+
+
+{- Note on optional fields:
+
+We do not actually care about whether 'Aeson.omitNothingFields' option is set
+or not, because Postgres treats "null" fields and absent fields equally - they
+are SQL nulls.
+
+For instance, this is good for our sorting, because fields which are "Maybe"
+can be justifiably sorted with nulls-first/nulls-last option specified.
+-}
```

Comment:
Better to move it to module description, so it appears in Haddock.

Reply:
This comment is more about an implementation detail - whether to consider `omitNothingFields` Aeson option. I can move it next to the related typeclass method implementation.